### PR TITLE
Remainder -> Modulo in accordance with Prism

### PR DESCRIPTION
**Pull request by Marckvdv**
Regarding issue #221. Correctly implements the modulo operator for expressions in accordance with Prism, by introducing an ITE for dealing with negative numbers. I added a test to check the behavior.

### CODE REVIEW

File: src/storm/storage/expressions/Expression.cpp [hunk 1]
```diff
@@ -300,9 +300,13 @@ Expression operator/(Expression const& first, Expression const& second) {
 
 Expression operator%(Expression const& first, Expression const& second) {
     assertSameManager(first.getBaseExpression(), second.getBaseExpression());
-    return Expression(std::shared_ptr<BaseExpression>(new BinaryNumericalFunctionExpression(
+    // First we compute the remainder
+    Expression remainder(std::shared_ptr<BaseExpression>(new BinaryNumericalFunctionExpression(
         first.getBaseExpression().getManager(), first.getType().modulo(second.getType()), first.getBaseExpressionPointer(), second.getBaseExpressionPointer(),
         BinaryNumericalFunctionExpression::OperatorType::Modulo)));
+
+    // Deal with negative `first` in accordance with Prism
```
Comment (tquatmann) [inline src/storm/storage/expressions/Expression.cpp:308]:
```suggestion
    // Deal with negative `first` in accordance with Prism and Jani
```
The Jani spec says:
> Unlike e.g. in most implementations of C, this operation computes the remainder of the Euclidean division: the result is always positive and x % 0 is division by zero.

As far as I can see, this is what we are doing now with this fix, right?

Comment (Marckvdv) [reply src/storm/storage/expressions/Expression.cpp:308]: Yes, except for the `x % 0` part. For normal integers that is going to give some divide/modulo zero error.

**Comment by tquatmann** [comment]
I just realized that `storm/storage/expressions/Expression.cpp` also implements a 
```
Expression modulo(Expression const& first, Expression const& second);
```
I guess this fix also has to be applied there?

**Comment by Marckvdv** [comment]
Looking at the code, the function body of `modulo` can simply be replaced by `return first % second;`.

Alternatively we can decide to put the actual code in `modulo` and replace the function body of `operator%` with `return modulo(first, second)`.

**Comment by tquatmann** [comment]
I created a prism file with a *ridiculous* amount of (nested!) modulos in there to check how much this can affect performance.
The model checking time went from `5s` (without the fix) to `15s` (with the fix). My suggestion above did not yield any improvement on this particular model.

Modulo does not seem to occur very often and my example is very artificial, so I don't object merging this as a simple fix to #221. But IMO we should create an issue  for future reference.

Here's my prism file which I built with `-const N=4,Q=4`: [polling.txt](https://github.com/moves-rwth/storm/files/8770893/polling.txt)

**Comment by tquatmann** [comment]
> Looking at the code, the function body of `modulo` can simply be replaced by `return first % second;`.
> 
> Alternatively we can decide to put the actual code in `modulo` and replace the function body of `operator%` with `return modulo(first, second)`.

Good point! I think it's fine either way.

**Comment by Marckvdv** [comment]
Yeah it is too bad that there is no separation of modulo/remainder in ExprTk... Some alternatives could be:
* `(n + (i % n)) % n`: branchless but probably much slower
* `i - floor(i/n)*n`: implementing modulo manually

**Pull request merged**